app/grandchallenge/hanging_protocols/static/js/hanging_protocols/hanging_protocol_visualization.js

```diff
@@ -0,0 +1,90 @@
+function insertAfter(newNode, existingNode) {
+    existingNode.parentNode.insertBefore(newNode, existingNode.nextSibling);
+}
+
+function removeAllChildNodes(parent) {
+    while (parent.firstChild) {
+        parent.removeChild(parent.firstChild);
+    }
+}
+
+function showOrHideVisualizationDiv(divId, jsonString) {
+    if (jsonString != '' && jsonString != 'null') {
+            document.getElementById(divId).style.display = 'block';
+        } else {
+            document.getElementById(divId).style.display = 'none';
+    }
+}
+
+function getGridDimensions(json) {
+    let allX = [];
+    let allY = [];
+    let allW = [];
+    let allH = [];
+
+    for (let i = 0; i < json.length; i++) {
+        json[i].x == undefined ? undefined : allX.push(json[i].x);
+        json[i].y == undefined ? undefined : allY.push(json[i].y);
+        json[i].w == undefined ? undefined : allW.push(json[i].w);
+        json[i].h == undefined ? undefined : allH.push(json[i].h);
+    }
+    let totalWidth = (allX.length != json.length | allW.length != json.length) ? json.length : Math.max(...allX.map(function(num, idx) {return num+allW[idx]}));

```

Comment:
The `|` in the first part of the ternary operator here is a bitwise OR. It think you meant to use the logical OR (=`||` in JS `or` in Python). Functionally it will not make a difference in this case since you apply it on booleans (except for very minor performance increase due to short circuiting) but you should always use the logical OR in cases like these.


Also, the logic on these lines will fail silently if there are multiple viewports defined that only contain some of the `x`, `y`, `w` or `h` fields. I think it is probably best to add an extra requirement to the hanging protocol that either all of these fields for all viewports are filled in, or none of them. That way it will be a lot easier to calculate max heigh/width and it will be a lot clearer also for the users. Then this whole function can also be simplified.